utils/sub8_ros_tools/so3tools/demo_so3tools.py

```diff
@@ -0,0 +1,220 @@
+#    when the result is applied to some vector, we want to first apply nextbody_to_lastbody and
+#    then apply lastbody_to_world. Recall so3.plus is not commutative.
+# 6: Let lasbody be the current body, and recall that nextbody_to_lastbody = dq.
+#    Thus qnext = so3.plus(dq, qlast) = so3.plus(dq, q).
+# 7: Finally, set q = qnext, and repeat for the next timestep.
+# Unless it is designed to work on SO3, typical ODE solvers will try to use regular Euclidean addition
+# when integrating. If you just tell it  qdot = f(w,dt) = so3.quaternion_from_axle(w*dt)/dt, even though
+# that is correct, it will integrate by doing something analogous to  qnext = q + qdot*dt  (even RK4 is just
+# a more involved version of that). Since the + used Euclidean, the result will be fundamentally wrong.
+# To use a typical ODE solver, we would need to reformulate the quaternion state ODE into an approximate form where
+# addition is possible, like the one given in http://ssl.mit.edu/spheres/library/ASS2011_11-033_spheres.pdf. On each
+# iteration, the quaternion state will become less and less unit norm both due to computer roundoff (which plagues
+# every method) AND due to local approximation of SO3 as Euclidean. This is because the euclidean sum of two unit quaternions
+# is not a unit quaternion unless one of them represents an infinitesmal angle. That is, a hypothetical analytical solution
+# to their equation would be exactly correct, but a numerical solution will be less accurate than a numerical solution using
+# the angle-axis approach, which was explained above for the first order integration case. An equivalent but even more numerically
+# accurate RK4 implementation of the angle axis method could and should be implemented.
+# ---
+# Simulate over t_arr:
+for i,t in enumerate(t_arr):
+    # Record current state:
+    q_history[i,:] = np.copy(q)
+    roll_history[i], pitch_history[i], yaw_history[i] = trns.euler_from_quaternion(q,'rxyz')
+    body_world_history[:,:,i] = so3.apply(q, body)
+    # Current values needed to compute next state:
+    I_world = so3.apply(q, so3.apply(q, I).T)  # current inertia matrix in world frame noting I = I.T
+    H = I_world.dot(w)  # current angular momentum
+    dq = so3.quaternion_from_axle(w*dt)  # change in orientation for this timestep instant
+    # PD controller... in ONLY THREE LINES BITCHES:
+    q_err = so3.error(q, q_des)  # q_err is an axle
+    w_err = w_des - w
+    torque = kp*q_err + kd*w_err
+    # Compute next state:
+    q = so3.plus(dq, q)  # new orientation computed using dq and old q
+    I_world = so3.apply(q, so3.apply(q, I).T)  # new I_world computed using new q
+    H = H + torque*dt  # new H from old H and torque
+    w = npl.inv(I_world).dot(H)  # new angular velocity computed using new I_world and new H
+
+################################################# DISPLAY
+
+fig = plt.figure()
+fig.suptitle('Orientation State Evolution', fontsize=24)
+
+if animate == True:
```

Comment:
if animate is True:

is bad. I think pylint allows it so that there is some way of checking that a value is literally exactly `True`, but you'd pretty much never want to actually do that.

Use the first example.